File: sdk/java/tests/ClientSanityRunner.kt
import payments.*
import java.io.File
import java.nio.charset.StandardCharsets
import org.json.JSONObject
import java.util.Base64
import java.util.Scanner

fun main() {
    val sc = Scanner(System.`in`)
    val sb = StringBuilder()
    while (sc.hasNextLine()) {
        sb.append(sc.nextLine())
    }
    
    val input = JSONObject(sb.toString())
    val scenarioId = input.getString("scenario_id")
    val sourceId = input.getString("source_id")
    val reqData = input.getJSONObject("request")
    val proxy = input.optJSONObject("proxy")
    val clientTimeoutMs = input.optLong("client_timeout_ms", -1L)
    val clientResponseTimeoutMs = input.optLong("client_response_timeout_ms", -1L)

    val headers = mutableMapOf<String, String>()
    if (reqData.has("headers")) {
        val h = reqData.getJSONObject("headers")
        for (key in h.keys()) {
            headers[key] = h.getString(key)
        }
    }
    headers["x-source"] = sourceId
    headers["x-scenario-id"] = scenarioId

    val body = if (reqData.has("body") && !reqData.isNull("body")) {
        val b = reqData.getString("body")
        if (b.startsWith("base64:")) {
            Base64.getDecoder().decode(b.substring(7))
        } else {
            b.toByteArray(StandardCharsets.UTF_8)
        }
    } else null

    val request = HttpRequest(
        url = reqData.getString("url"),
        method = reqData.getString("method"),
        headers = headers,
        body = body
    )

    val httpConfig = when {
        clientTimeoutMs > 0 -> HttpConfig.newBuilder().setTotalTimeoutMs(clientTimeoutMs.toInt()).build()
        clientResponseTimeoutMs > 0 -> HttpConfig.newBuilder().setResponseTimeoutMs(clientResponseTimeoutMs.toInt()).build()
        else -> null
    }

    val output = JSONObject()
    try {
        val clientConfig = if (proxy != null) {
            val httpUrl = proxy.optString("http_url", "")
            if (httpUrl.isNotEmpty()) {
                HttpConfig.newBuilder()
                    .setProxy(ProxyOptions.newBuilder().setHttpUrl(httpUrl).build())
                    .build()
            } else null
        } else null

        val client = HttpClient.createClient(clientConfig)
        val sdkResponse = HttpClient.execute(request, httpConfig, client)
        
        val ct = (sdkResponse.headers["content-type"] ?: "").lowercase()
        val bodyStr = if ("application/octet-stream" in ct) {
            Base64.getEncoder().encodeToString(sdkResponse.body)
        } else {
            String(sdkResponse.body, StandardCharsets.UTF_8)
        }
        
        output.put("response", JSONObject().apply {
            put("statusCode", sdkResponse.statusCode)
            put("headers", JSONObject(sdkResponse.headers))
            put("body", bodyStr)
        })
    } catch (e: Exception) {
        val code = if (e is NetworkError) e.code.name else "UNKNOWN_ERROR"
        output.put("error", JSONObject().apply {
            put("code", code)
            put("message", e.message ?: e.toString())
        })
    }

    println(output.toString())
}

File: sdk/rust/src/bin/client_sanity_runner.rs
use base64::{engine::general_purpose, Engine as _};
use common_utils::request::Method;
use hyperswitch_payments_client::http_client::{HttpClient, HttpOptions, HttpRequest, ProxyConfig};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::io::{self, Read};

#[derive(Deserialize)]
struct RunnerInput {
    scenario_id: String,
    source_id: String,
    request: RequestDetails,
    client_timeout_ms: Option<u64>,
    client_response_timeout_ms: Option<u64>,
    proxy: Option<ScenarioProxy>,
}

#[derive(Deserialize)]
struct ScenarioProxy {
    http_url: Option<String>,
    https_url: Option<String>,
}

#[derive(Deserialize)]
struct RequestDetails {
    url: String,
    method: String,
    headers: HashMap<String, String>,
    body: Option<String>,
}

#[derive(Serialize)]
struct RunnerOutput {
    #[serde(skip_serializing_if = "Option::is_none")]
    response: Option<SdkResponse>,
    #[serde(skip_serializing_if = "Option::is_none")]
    error: Option<SdkError>,
}

#[derive(Serialize)]
struct SdkResponse {
    #[serde(rename = "statusCode")]
    status_code: u16,
    headers: HashMap<String, String>,
    body: String,
}

#[derive(Serialize)]
struct SdkError {
    code: String,
    message: String,
}

fn output_json(out: &RunnerOutput) {
    if let Ok(s) = serde_json::to_string(out) {
        println!("{}", s);
    }
}

#[tokio::main]
async fn main() {
    let output = run().await;
    output_json(&output);
}

async fn run() -> RunnerOutput {
    let mut buffer = String::new();
    if io::stdin().read_to_string(&mut buffer).is_err() {
        return RunnerOutput {
            response: None,
            error: Some(SdkError {
                code: "RUNNER_CRASH".to_string(),
                message: "Failed to read stdin".to_string(),
            }),
        };
    }

    let input: RunnerInput = match serde_json::from_str(&buffer) {
        Ok(i) => i,
        Err(e) => {
            return RunnerOutput {
                response: None,
                error: Some(SdkError {
                    code: "RUNNER_CRASH".to_string(),
                    message: format!("Invalid input JSON: {}", e),
                }),
            };
        }
    };

    let method = match input.request.method.as_str() {
        "POST" => Method::Post,
        "GET" => Method::Get,
        "PUT" => Method::Put,
        "DELETE" => Method::Delete,
        "PATCH" => Method::Patch,
        _ => {
            return RunnerOutput {
                response: None,
                error: Some(SdkError {
                    code: "RUNNER_CRASH".to_string(),
                    message: format!("Unsupported method: {}", input.request.method),
                }),
            };
        }
    };

    let mut headers = input.request.headers.clone();
    headers.insert("x-source".to_string(), input.source_id);
    headers.insert("x-scenario-id".to_string(), input.scenario_id);

    let body = input.request.body.map(|b| {
        if b.starts_with("base64:") {
            general_purpose::STANDARD
                .decode(b.trim_start_matches("base64:"))
                .unwrap_or_default()
        } else {
            b.into_bytes()
        }
    });

    let request = HttpRequest {
        url: input.request.url.clone(),
        method,
        headers,
        body,
    };

    let proxy_cfg = input.proxy.as_ref().map(|p| ProxyConfig {
        http_url: p.http_url.clone(),
        https_url: p.https_url.clone(),
        bypass_urls: vec![input.request.url.clone()],
    });

    let options = HttpOptions {
        proxy: proxy_cfg,
        total_timeout_ms: input.client_timeout_ms.map(|ms| ms as u32),
        response_timeout_ms: input.client_response_timeout_ms.map(|ms| ms as u32),
        ..Default::default()
    };

    let client = match HttpClient::new(options) {
        Ok(c) => c,
        Err(e) => {
            return RunnerOutput {
                response: None,
                error: Some(SdkError {
                    code: e.error_code().to_string(),
                    message: e.to_string(),
                }),
            };
        }
    };

    let sdk_result = client.execute(request, None).await;

    match sdk_result {
        Ok(resp) => {
            let ct = resp
                .headers
                .get("content-type")
                .map(|s| s.to_lowercase())
                .unwrap_or_default();
            let body_str = if ct.contains("application/octet-stream") {
                general_purpose::STANDARD.encode(&resp.body)
            } else {
                String::from_utf8_lossy(&resp.body).into_owned()
            };
            RunnerOutput {
                response: Some(SdkResponse {
                    status_code: resp.status_code,
                    headers: resp.headers,
                    body: body_str,
                }),
                error: None,
            }
        }
        Err(e) => RunnerOutput {
            response: None,
            error: Some(SdkError {
                code: e.error_code().to_string(),
                message: e.to_string(),
            }),
        },
    }
}
